### DISCUSSION TITLE
Как отправить события на определённый канал

### DISCUSSION THREAD

**Discussion by laika65** [melanchall/drywetmidi#166] [Q&A] [answered]
Имею простой метод, но не понимаю как отправить  на определённый канал.
```
using (var outputDevice = OutputDevice.GetByName("MIDI Device"))
{
        outputDevice.SendEvent(new NoteOnEvent());
        outputDevice.SendEvent(new NoteOffEvent());
}
```
И ещё, почему то при проигрывание через "Microsoft GS Wavetable Synth" нет звука и в консоли куча эксепшионов, в чём может быть причина?

**Comment by melanchall** [comment, marked as answer]
Все так называемые "канальные" события (коими являются в том числе `NoteOnEvent` и `NoteOffEvent`) имеют свойство `Channel`. Установите его в нужное значение. У библиотеки есть полное руководство по всему API, пожалуйста, изучите его: https://melanchall.github.io/drywetmidi.

А звука, конечно же, не будет. Вы отправляете событие нажатия ноты и сразу же событие отпускания ноты. Длительность ноты тут 0. Либо вставьте паузу между событиями, либо используйте класс `Playback` (см. документацию опять же).

Кроме того, вы отправляете последнее событие, после чего ваш код покидает секцию `using`, в результате чего API по взаимодействию с устройством освобождается, и событие может быть вообще не проиграно. Нужно сохранять устрйоство в переменную или поле, и по ненадобности освобождать в конце.

Ну и по исключениям: я их не вижу, так что сказать ничего не могу. Покажите их, пожалуйста.

**Comment by laika65** [comment]
Извиняюсь, только начал знакомится и с языком и с вашей библиотекой и задаю вопросы раньше чем нужно. С каналами разобрался.

По поводу виндовского синта:
Вот код

```
var midiFile = MidiFile.Read(@"z:\MuZSrc\MIDI\Piano.mid");
using (var outputDevice = OutputDevice.GetByName("Microsoft GS Wavetable Synth"))
{
    using (var playback = midiFile.GetPlayback(outputDevice))
    {
        playback.Play();
    }
}
```
Эксептионы:
Exception thrown at 0x00007FFFF88E478C in Pad.exe: Microsoft C++ exception: wil::ResultException at memory location 0x000000282267DE60.
Exception thrown at 0x00007FFFF88E478C in Pad.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFFF88E478C in Pad.exe: Microsoft C++ exception: wil::ResultException at memory location 0x000000282267DF20.
Exception thrown at 0x00007FFFF88E478C in Pad.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFFF88E478C in Pad.exe: Microsoft C++ exception: wil::ResultException at memory location 0x000000282267DF90.
Exception thrown at 0x00007FFFF88E478C in Pad.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.

Виндовс 11, в 10-ой было также.

Тестовая миди
https://www72.zippyshare.com/v/icvjL0DG/file.html
Не знаю ещё как тут прикреплять.

**Reply by melanchall** [reply]
Проверил в Windows 10, никаких проблем. Прикреплять сюда можно обыкновенным drag'n'drop-ом. Прикрепите, пожалуйста, целиком солюшн (папку с sln-файлом).

**Reply by laika65** [reply]
Переделал под консольное приложение и заиграло. Проблема, видимо, в UWP/WinUI приложении. Вроде видел что у кого то были какие то проблемы с UWP. При использовании других портов таких проблем нет, только с майкросовтовским синтом ерунда.
